src/Microsoft.FeatureManagement/VariantServiceProvider.cs

```diff
@@ -0,0 +1,86 @@
+    {
+        private readonly IEnumerable<TService> _services;
+        private readonly IVariantFeatureManager _featureManager;
+        private readonly string _variantFeatureName;
+
+        /// <summary>
+        /// Creates a variant service provider.
+        /// </summary>
+        /// <param name="services">Implementation variants of TService.</param>
+        /// <param name="featureManager">Feature manager to get the assigned variant of the variant feature flag.</param>
+        /// <exception cref="ArgumentNullException">Thrown if <paramref name="services"/> is null.</exception>
+        /// <exception cref="ArgumentNullException">Thrown if <paramref name="featureManager"/> is null.</exception>
+        public VariantServiceProvider(IEnumerable<TService> services, IVariantFeatureManager featureManager)
+        {
+            _services = services ?? throw new ArgumentNullException(nameof(services));
+            _featureManager = featureManager ?? throw new ArgumentNullException(nameof(featureManager));
+        }
+
+        /// <summary>
+        /// The variant feature flag used to assign variants.
+        /// </summary>
+        public string VariantFeatureName
+        {
+            get => _variantFeatureName;
+
+            init
+            {
+                _variantFeatureName = value ?? throw new ArgumentNullException(nameof(value));
+            }
+        }
+
+        /// <summary>
+        /// Gets implementation of TService according to the assigned variant from the feature flag.
+        /// </summary>
+        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
+        /// <returns>An implementation matched with the assigned variant. If there is no matched implementation, it will return null.</returns>
+        public async ValueTask<TService> GetServiceAsync(CancellationToken cancellationToken)
+        {
+            Debug.Assert(_variantFeatureName != null);
+
+            Variant variant = await _featureManager.GetVariantAsync(_variantFeatureName, cancellationToken);
+
+            TService implementation = null;
+
+            if (variant != null)
+            {
+                implementation = _services.FirstOrDefault(service => 
+                    IsMatchingVariant(
+                        service.GetType(),
+                        variant));
+            }
+
+            return implementation;
+        }
+
+        private bool IsMatchingVariant(Type implementationType, Variant variant)
+        {
+            Debug.Assert(variant != null);
+
+            string implementationName = ((VariantServiceAliasAttribute)Attribute.GetCustomAttribute(implementationType, typeof(VariantServiceAliasAttribute)))?.Alias;
+
+            if (implementationName == null)
+            {
+                implementationName = implementationType.Name;
+            }
+
+            return string.Equals(implementationName, variant.Name, StringComparison.OrdinalIgnoreCase);
```

Comment:
Just to confirm, is it case insensitive for feature filter too?

Reply:
Yes. Feature filter name is case insensitive. https://github.com/microsoft/FeatureManagement-Dotnet/blob/main/src/Microsoft.FeatureManagement/FeatureManager.cs#L377